### PR TITLE
Rehaul DNS Error Handling

### LINKED ISSUE #994: DoH bug with no answer
# 🐛 Summary #
In some rare cases, a DNS request might not result in NXDOMAIN but still not return an answer section. One specific case where this happens is when a tenant's license is expired and you request the txt record for the ".onmicrosoft.com" domain, though that is by no means the only way that could happen. Our code for the traditional DNS request handles that scenario just fine but the DoH retry code tries to iterate through a null object and results in an exception.

## To reproduce ##
Run ScubaGear against an expire tenant while on a network that does not block DoH. Note the warning printed to command-line ("WARNING: Get-ScubaSpfRecord: for 1 domain(s), the tradtional DNS queries returned an empty answer section and the DoH queries failed..."

Side note: note the typo with "tradtional" in the command-line warning, might as well fix that as well.

## Expected behavior ##
No exception.

I expect the fix will look like this (ExportEXOProvider.psm1), but I haven't fully tested:
image

## Any helpful log output or screenshots ##
Log data in the provider output. You should see something like this under the "spf_records" section.
image

### LINKED ISSUE #1148: Invoke-RobustDnsTxt in some cases logs an incorrect number of responses
# 🐛 Summary #
When:
- Traditional DNS isn't working
- But DoH is working
- And the domain being resolved only returns 1 txt record

The function logs the length the txt record as the number of records returned.

Note that the HTML report is not impacted in any way by this bug. Only the log data inside the json output is impacted.

## Expected behavior ##
Just 1 txt record to be listed in the log.

## Any helpful log output or screenshots ##
To reproduce the error:

While on a network that has DoH available:
1. Modify the Invoke-RobustDnsTxt method to force DoH. For example: 
image
2. Modify Get-ScubaSpfRecord to add a known domain that has just one record. For example: 
image
3. Note the problem inside the json output:
image

### LINKED ISSUE #1357: Enhance details message for SPF
# 💡 Summary #
Add more details to the "Details" message for SPF.

## Motivation and context ##
The policy reads as:
> An SPF policy SHALL be published for each domain that fails all non-approved senders.

A SPF policy such as `v=spf1 include:_spf.google.com ~all` would actually not meet the baseline because it "soft fails" (i.e., `~all` instead of `-all`). But if you have an SPF policy like that, all ScubaGoggles says is `x out y agency domain(s) found in violation: example.com, ...`
If I were a user and I saw that message, I would likely be confused, "But I do have an SPF policy, why is it still failing in ScubaGoggles?" It would be nice if ScubaGoggles would tell me exactly why that domain failed the SPF requirements (e.g., "Domain x has an SPF policy but only soft-fails `~all` other domains.")

## Implementation notes ##
Maybe the simplest way to do this would be something like this:

```
The following domains have not published SPF polices:
- example1
- example2

The following domains have published SPF policies but do not fail (i.e. "-all") all non-approved senders:
- example3
- example4
```

## Acceptance criteria ##
- [ ] Details message is more specific

### LINKED ISSUE #1479: Revisit Invoke-RobustDnsTxt error handling
# 💡 Summary #
An exception thrown while resolving one domain shouldn't prevent the other domains from being tested. There are some cases where that can happen with the current implementation of Invoke-RobustDnsTxt.

## Motivation and context ##
Invoke-RobustDnsTxt handles the following cases:
- Non-existent domain names (NXDOMAIN)
- Empty answer sections
- One-time network failures
- Split DNS scenarios where a domain isn't accessible locally but can be resolved on the public Internet (assuming DoH isn't blocked).

But if an exception occurs outside of that scope, `Invoke-RobustDnsTxt` will throw an exception. For example, if both `Resolve-DnsName` and `Invoke-WebRequest` both repeatedly timeout (e.g., due to a firewall), an exception will be thrown. For example, a misconfigured authoritative server might return a SERVFAIL for one domain, but that wouldn't imply that other domains would fail, so that shouldn't prevent ScubaGear from querying the other domains.

Resulting issues:
- https://github.com/cisagov/ScubaGear/issues/1477
- https://github.com/cisagov/ScubaGear/issues/1478

## Implementation notes ##
`Invoke-RobustDnsTxt` ends with an if/elseif/else.
- `if`: We got our answer, no change needed for this case.
- `elseif`: We got an empty answer section, experience has shown that this might be due to a split DNS setup. A warning is printed by the calling function in this case.
- `else`: This is the case that needs to be changed. Simplest thing would be to convert that `throw` to `Write-Warning`, then have the function return `@{"Answers" = $Answers; "HighConfidence" = $true; "LogEntries" = $LogEntries}`. Main problem with that strategy is that does introduce the possibility of a log of long warnings to be printed. Maybe that's ok?

## Acceptance criteria ##
An error resolving one domain name doesn't prevent the rest of the domains from being attempted.

### LINKED ISSUE #1510: Get-ScubaSpfRecord returns in rdata all TXT records not only SPF
# 🐛 Summary #

Get-ScubaSpfRecord returns in rdata all TXT records not only SPF

## To reproduce ##

Check on domains with TXT records, but not SPF.

## Expected behavior ##

Only SPF records to be present in rdata

### LINKED ISSUE #1534: Ensure long DNS text records aren't split
# 🐛 Summary #
Long DNS records are being split, causing false positives for the SPF evaluation.

## To reproduce ##
See https://github.com/cisagov/ScubaGear/issues/1510#issuecomment-2600888497.

## Expected behavior ##
The SPF records to be properly read.

## Any helpful log output or screenshots ##
See https://github.com/cisagov/ScubaGear/issues/1510#issuecomment-2600888497.

### PR DESCRIPTION
**Pull request by adhilto**
#  #
  

## 🗣 Description ##
- Corrected various DNS-related bugs
- Abandoned old error complicated handling method that involved tracking if an answer was "high confidence"
- Split up old unwieldy `Invoke-RobustDnsTxt` function into 3 functions. 
- Don't print warning on empty answer. Warning was confusing and not very helpful. Instead, added table with all DNS logs at the end of EXO report. That way users can see exactly how ScubaGear reached it's conclusions and identify any unexpected results themselves (e.g., due to split horizon).

The DNS log table may be a good place to point users to the DNS configuration options (#1533) once we implement that.

## 💭 Motivation and context ##
Closes #1479.
Closes #1510.
Closes #1357.
Closes #1148.
Closes #994.
Closes #1534.

## 🧪 Testing ##
- Updated and added new unit tests
- Manually tested against both E5 and G5
- Temporarily modified the `Get-ScubaSpfRecord` function to add some edge cases
image

## 📷 Screenshots ##
image

image

## ✅ Pre-approval checklist ##

- [x] This PR has an informative and human-readable title.
- [x] PR targets the correct parent branch (e.g., main or release-name) for merge.
- [x] Changes are limited to a single goal - *eschew scope creep!*
- [x] Changes are sized such that they do not touch excessive number of files.
- [x] *All* future TODOs are captured in issues, which are referenced in code comments.
- [x] These code changes follow the ScubaGear [content style guide](https://github.com/cisagov/ScubaGear/blob/main/CONTENTSTYLEGUIDE.md).
- [x] Related issues these changes resolve are linked preferably via [closing keywords](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword).
- [x] All relevant type-of-change labels added.
- [x] All relevant project fields are set.
- [x] All relevant repo and/or project documentation updated to reflect these changes.
- [x] Unit tests added/updated to cover PowerShell and Rego changes.
- [x] Functional tests added/updated to cover PowerShell and Rego changes.
- [x] All relevant functional tests passed.
- [x] All automated checks (e.g., linting, static analysis, unit/smoke tests) passed.

## ✅ Pre-merge checklist ##

- [ ] PR passed smoke test check.
- [ ] Feature branch has been rebased against changes from parent branch, as needed

  Use `Rebase branch` button below or use [this](https://www.digitalocean.com/community/tutorials/how-to-rebase-and-update-a-pull-request) reference to rebase from the command line.
- [ ] Resolved all merge conflicts on branch
- [ ] Notified merge coordinator that PR is ready for merge via comment mention
- [ ] Demonstrate changes to the team for questions and comments. 
    (Note: Only required for issues of size `Medium` or larger)

## ✅ Post-merge checklist ##

- [ ] Feature branch deleted after merge to clean up repository.
- [ ] Verified that all checks pass on parent branch (e.g., main or release-name) after merge.

### CODE REVIEW

File: PowerShell/ScubaGear/Modules/Providers/ExportEXOProvider.psm1 [hunk 1]
```diff
@@ -238,119 +293,165 @@ function Invoke-RobustDnsTxt {
...
+
+    if ($script:DohServer -eq "") {
+        # We haven't determined if DoH is available yet, select the first server that works
+        $script:DohServer = Select-DohServer
+    }
+    if ($null -eq $script:DohServer) {
+        # None of the DoH servers are accessible
+        $LogEntries += @{"query_name"=$Qname; "query_method"="DoH"; "query_result"="NA, DoH servers unreachable"}
+    }
+    else {
```
Comment (buidav) [inline PowerShell/ScubaGear/Modules/Providers/ExportEXOProvider.psm1]:
```suggestion
    if ($null -eq $script:DohServer) {
        # None of the DoH servers are accessible
        $LogEntries += @{"query_name"=$Qname; "query_method"="DoH"; "query_result"="NA, DoH servers unreachable"}
         return @{
            "Answers" = $Answers;
            "NXDomain" = $NXDomain;
            "LogEntries" = $LogEntries;
            "Errors" = $Errors
        }
    }
```
Do not take the suggestion as is. It's just an example.
Let's duplicate the return statement here and have an early exit of the function. 
This allows us to take the main logic of the function out of  the `else` statement. 
This greatly helps with the readability of the function.

File: PowerShell/ScubaGear/Modules/Providers/ExportEXOProvider.psm1 [hunk 2]
```diff
@@ -238,119 +293,165 @@ function Invoke-RobustDnsTxt {
...
-                    elseif ($ResponseBody.Status -eq 3) {
-                        # 3 indicates NXDomain. The DNS query succeeded, but the domain did not exist.
-                        # Set $Success to $true, because event though the domain does not exist, the
-                        # query succeeded, and this came from an external resolver so split horizon is
-                        # not an issue here.
+    @{
+        "Answers" = $Answers;
+        "NXDomain" = $NXDomain;
+        "LogEntries" = $LogEntries;
+        "Errors" = $Errors
```
Comment (buidav) [inline PowerShell/ScubaGear/Modules/Providers/ExportEXOProvider.psm1]:
```suggestion
        "Answers" = $Answers;
        "NXDomain" = $NXDomain;
        "LogEntries" = $LogEntries;
        "Errors" = $Errors;
```

**Comment by adhilto** [comment]
I added code to truncate the tables for tenants with a log of domains.

image
Screenshot 2025-06-16 091245

Tested with this:
Screenshot 2025-06-16 091632
